File: apply_force_pkg/src/apply_force.cpp
#include <ros/ros.h>
#include <gazebo_msgs/ApplyBodyWrench.h>
#include <geometry_msgs/Wrench.h>
#include <geometry_msgs/Vector3.h>
#include <math.h>

void applyForceInXDirection(const std::string& body_name, double force_magnitude) {
    ros::NodeHandle n;
    ros::ServiceClient client = n.serviceClient<gazebo_msgs::ApplyBodyWrench>("/gazebo/apply_body_wrench");
    gazebo_msgs::ApplyBodyWrench srv;

    srv.request.body_name = body_name;
    srv.request.reference_frame = body_name;

    geometry_msgs::Wrench wrench;
    wrench.force.x = force_magnitude;
    wrench.force.y = 0;
    wrench.force.z = 0;
    srv.request.wrench = wrench;

    srv.request.start_time = ros::Time(0);
    srv.request.duration = ros::Duration(0.1);

    if (client.call(srv)) {
        ROS_INFO("Successfully applied force: %f N in X direction", force_magnitude);
    } else {
        ROS_ERROR("Failed to call service apply_body_wrench");
    }
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "apply_varying_force_in_x");
    ros::NodeHandle n;

    std::string body_name = "qm::ft_sensor";  // Replace with your sensor link name
    double max_force = 55.0;  // Maximum force magnitude in Newtons
    double min_force = -35.0; // Minimum force magnitude in Newtons
    double force_step = 5;  // Step size for each iteration
    double current_force = min_force;
    bool increasing = true;

    ros::Rate rate(10);  // 10 Hz update rate

    while (ros::ok()) {
        applyForceInXDirection(body_name, current_force);

        if (increasing) {
            current_force += force_step;
            if (current_force >= max_force) {
                current_force = max_force;
                increasing = false;
            }
        } else {
            current_force -= force_step;
            if (current_force <= min_force) {
                current_force = min_force;
                increasing = true;
            }
        }

        rate.sleep();
    }

    return 0;
}

File: apply_force.py
import rospy
import math
from geometry_msgs.msg import Wrench, Vector3
from gazebo_msgs.srv import ApplyBodyWrench, ApplyBodyWrenchRequest
import time

def apply_force_at_angle(body_name, force_magnitude, angle):
    """
    Apply a force to a specified body at a given angle around the Z-axis.
    :param body_name: The name of the body to apply the force to.
    :param force_magnitude: The magnitude of the force to apply.
    :param angle: The angle in radians at which to apply the force, measured from the Y-axis.
    """
    rospy.wait_for_service('/gazebo/apply_body_wrench')
    try:
        apply_body_wrench = rospy.ServiceProxy('/gazebo/apply_body_wrench', ApplyBodyWrench)
        wrench = Wrench()
        wrench.force.x = 0
        wrench.force.y = force_magnitude * math.cos(angle)
        wrench.force.z = force_magnitude * math.sin(angle)
        request = ApplyBodyWrenchRequest()
        request.body_name = body_name
        request.reference_frame = body_name
        request.wrench = wrench
        request.start_time = rospy.Time(0)
        request.duration = rospy.Duration(0.1)
        apply_body_wrench(request)
    except rospy.ServiceException as e:
        rospy.logerr("Service call failed: %s" % e)

def main():
    rospy.init_node('apply_varying_force')
    body_name = "qm::ft_sensor"  # Replace with your sensor link name
    force_magnitude = 30.0  # Force magnitude in Newtons
    start_angle = -math.pi / 2  # Start from -90 degrees (Y-axis negative direction)
    end_angle = math.pi / 2  # End at 90 degrees (Y-axis positive direction)
    num_steps = 100  # Number of steps to transition the force direction

    step_angle = (end_angle - start_angle) / num_steps
    rate = rospy.Rate(10)  # 10 Hz update rate

    for step in range(num_steps + 1):
        current_angle = start_angle + step * step_angle
        apply_force_at_angle(body_name, force_magnitude, current_angle)
        rate.sleep()

if __name__ == "__main__":
    main()
